include/ignition/math/SignalStats.hh

```diff
@@ -66,17 +58,13 @@ namespace ignition
       /// \brief Forget all previous data.
       public: virtual void Reset();
 
-#ifdef _WIN32
-// Disable warning C4251 which is triggered by
-// std::unique_ptr
-#pragma warning(push)
-#pragma warning(disable: 4251)
-#endif
+      public: class Implementation;
+      IGN_UTILS_WARN_IGNORE__DLL_INTERFACE_MISSING
       /// \brief Pointer to private data.
-      protected: std::unique_ptr<SignalStatisticPrivate> dataPtr;
-#ifdef _WIN32
-#pragma warning(pop)
-#endif
+      /// Not using macros as it needs to be protected to be accessed from
+      /// derived classes
+      protected: ::ignition::utils::ImplPtr<Implementation> dataPtr;
+      IGN_UTILS_WARN_RESUME__DLL_INTERFACE_MISSING
```

Comment:
I don't think it defeats the purpose.  The `ImplPtr` mainly serves to ensure that correct copy/move/assignment operators are created when using the PIMPL pattern.  Getting the DLL_INTERFACE macros is kind of a bonus.

I think that the biggest oversight may be making the macros hardcoded to `private:`, where we could leave it up to the implementor?

Reply:
I believe the original purpose of the PIMPL pattern is to hide the implementation so that changes can be made without breaking ABI. Since protected members are meant to be visible by derived classes, it doesn't make sense to PIMPLize them ([Herb Sutter thinks so too](http://www.gotw.ca/publications/mill05.htm)).

But it sounds like `ImplPtr` is being used here with a different goal. It isn't for PIMPL but for achieving the rule-of-5. In that case, isn't it simpler to just have the protected members directly in the class instead or wrapping them into a `dataPtr`?

Reply:
I think a better approach would be to add `protected` setters and getters for the `data`, `extraData`, and `count` member variables from [SignalStatsPrivate.hh](https://github.com/ignitionrobotics/ign-math/blob/ign-math6/src/SignalStatsPrivate.hh#L34-L41). Then it could follow a more typical PIMPL pattern.

Reply:
> I believe the original purpose of the PIMPL pattern is to hide the implementation so that changes can be made without breaking ABI. Since protected members are meant to be visible by derived classes, it doesn't make sense to PIMPLize them (Herb Sutter thinks so too).

Sorry, I was thinking you were objecting to the DLL interface macros. 

In that case I agree, it makes sense to either a) drop PIMPL here, or b) expose the methods as @scpeters says. I'm happy to go ahead and fix this.  Any preference?

Reply:
+1 to the getters and setters, that was my original suggestion above. I think it can come in a separate PR though, and we leave this class out of this one for now, what do you think?

Reply:
sure, you can leave `SignalStats.hh` unchanged in this PR and fix it separately

Reply:
`SignalStats.hh` has been reverted.